controllers/azuremachine_reconciler_test.go

```diff
@@ -0,0 +1,136 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"testing"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api-provider-azure/api/v1alpha2"
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+	"sigs.k8s.io/cluster-api-provider-azure/cloud/scope"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+func TestGetControlPlaneMachines(t *testing.T) {
+	controlPlaneMachine := clusterv1.Machine{
+		ObjectMeta: v1.ObjectMeta{
+			Name:   "control-plane-0",
+			Labels: map[string]string{},
+		},
+	}
+	controlPlaneMachine.ObjectMeta.Labels[clusterv1.MachineControlPlaneLabelName] = "cp-0"
+
+	agentMachine := clusterv1.Machine{
+		ObjectMeta: v1.ObjectMeta{
+			Name:   "machine-0",
+			Labels: map[string]string{},
+		},
+	}
+
+	cases := []struct {
+		name        string
+		machineList *clusterv1.MachineList
+		expected    []*clusterv1.Machine
+	}{
+		{
+			name: "empty",
+			machineList: &clusterv1.MachineList{
+				Items: []clusterv1.Machine{},
+			},
+			expected: []*clusterv1.Machine{},
+		},
+		{
+			name: "all",
+			machineList: &clusterv1.MachineList{
+				Items: []clusterv1.Machine{
+					controlPlaneMachine,
+				},
+			},
+			expected: []*clusterv1.Machine{&controlPlaneMachine},
+		},
+		{
+			name: "none",
+			machineList: &clusterv1.MachineList{
+				Items: []clusterv1.Machine{agentMachine},
+			},
+			expected: []*clusterv1.Machine{},
+		},
+		{
+			name: "some",
+			machineList: &clusterv1.MachineList{
+				Items: []clusterv1.Machine{agentMachine, controlPlaneMachine},
+			},
+			expected: []*clusterv1.Machine{&controlPlaneMachine},
+		},
+	}
+
+	for _, c := range cases {
+		c := c
+		t.Run(c.name, func(t *testing.T) {
+			t.Parallel()
+			actual := GetControlPlaneMachines(c.machineList)
+			if len(actual) != len(c.expected) {
+				t.Fatalf("Got unexpected GetControlPlaneMachines result. Expected length: %v. Got: %v.", len(actual), len(c.expected))
+			}
+			for i, v := range c.expected {
+				if v.ObjectMeta.Name != actual[i].ObjectMeta.Name {
+					t.Fatalf("Got unexpected GetControlPlaneMachines result. Expected Item: %v. Got: %v.", v, actual[i])
+
+				}
+			}
+		})
+	}
+}
+
+func TestIsAvailabilityZoneSupported(t *testing.T) {
+	s := azureMachineService{
+		machineScope: &scope.MachineScope{
+			Logger: log.Log.Logger,
+			AzureCluster: &v1alpha2.AzureCluster{
+				Spec: v1alpha2.AzureClusterSpec{
+					Location: "",
+				},
+			},
+		},
+	}
+
+	for _, l := range azure.SupportedAvailabilityZoneLocations {
+		s.machineScope.AzureCluster.Spec.Location = l
+		if s.isAvailabilityZoneSupported() != true {
+			t.Errorf("isAvailabilityZoneSupported should return true for supported region %s but returned %t", l, s.isAvailabilityZoneSupported())
+		}
+	}
+
+	unSupportedLocations := []string{
```

Comment:
Seems like an annoying test case, once somebody updates the list of supported, they need to update this unsupported set as well.

Recommend testing some random location name than a valid one. Example "randomregion" or something like that.

Reply:
yep good point. Changed.